[libc][stdfix] Add integer square root with fixed point output functions.

lntue opened the pull request:
Fix https://github.com/llvm/llvm-project/issues/83924.

lntue committed:
[libc][stdfix] Add integer square root with fixed point output functions.

nickdesaulniers commented:
> Even though it's not part of ISO 18037

What does that mean? Are these now functions outside of that spec? :unamused:

lntue commented:
> > Even though it's not part of ISO 18037
> 
> What does that mean? Are these now functions outside of that spec? 😒

Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.

Ideally, I would like to have all C23 math functions available for all numeric types.  We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?

nickdesaulniers commented:
> > > Even though it's not part of ISO 18037
> > 
> > 
> > What does that mean? Are these now functions outside of that spec? 😒
> 
> Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.
> 
> Ideally, I would like to have all C23 math functions available for all numeric types. We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?

Can we make that more explicit on https://libc.llvm.org/math/stdfix.html?

michaelrj-google commented on libc/spec/stdc_ext.td:
```diff
@@ -55,6 +55,9 @@ def StdcExt : StandardSpec<"stdc_ext"> {
           GuardedFunctionSpec<"sqrtuk", RetValSpec<UnsignedAccumType>, [ArgSpec<UnsignedAccumType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
           GuardedFunctionSpec<"sqrtulk", RetValSpec<UnsignedLongAccumType>, [ArgSpec<UnsignedLongAccumType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
+
+          GuardedFunctionSpec<"uhksqrtus", RetValSpec<UnsignedShortAccumType>, [ArgSpec<UnsignedShortType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
```
is this actually standardized or should this go in `llvm_libc_ext.td`?

nickdesaulniers replied:
right, I would prefer stronger language / guarantees between "implemented against a spec" vs straight up "extension."

both in documentation and in our headergen files as @michaelrj-google suggested.

lntue replied:
I've moved the specs of these functions to `llvm_libc_ext.td`.

michaelrj-google commented on libc/test/src/stdfix/ISqrtTest.h:
```diff
@@ -0,0 +1,66 @@
+  static constexpr OutType eps = FXRep::EPS();
+
+public:
+  typedef OutType (*SqrtFunc)(T);
```
do these tests properly handle the fast math case?

michaelrj-google reviewed (approved):
Overall LGTM with some nits

lntue committed:
Address comments.

lntue commented:
> > > > Even though it's not part of ISO 18037
> > > 
> > > 
> > > What does that mean? Are these now functions outside of that spec? 😒
> > 
> > 
> > Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.
> > Ideally, I would like to have all C23 math functions available for all numeric types. We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?
> 
> Can we make that more explicit on https://libc.llvm.org/math/stdfix.html?

I've updated stdfix's documentations.

michaelrj-google commented on libc/src/stdfix/CMakeLists.txt:
```diff
@@ -46,3 +46,29 @@ foreach(suffix IN ITEMS hr r lr hk k lk uhr ur ulr uhk uk ulk)
+    uhksqrtus.cpp
+  COMPILE_OPTIONS
+    -O3
+    -ffixed-point
```
if `LIBC_COMPILER_HAS_FIXED_POINT` is specified, then `-ffixed-point` is automatically added to all of the compile options. You don't have to additionally specify it here.

lntue replied:
Removed all redundant `-ffixed-point` flags.

lntue committed:
Remove redundant -ffixed-point flags.

nickdesaulniers commented on libc/docs/math/stdfix.rst:
```diff
@@ -93,6 +98,9 @@ roundfx
 =====================
 
+The following math functions are modeled after C99/C23 math functions for
+floating point types.
```
```diff
- floating point types.
+ floating point types, but are not part of the the ISO/IEC TR 18037:2008 spec.
```

lntue replied:
Done.

lntue committed:
Update docs.

lntue committed:
Clean up test logging.

nickdesaulniers commented on libc/spec/llvm_libc_ext.td:
```diff
@@ -51,9 +51,29 @@ def LLVMLibcExt : StandardSpec<"llvm_libc_ext"> {
-    Sched,
     Assert,
+    Sched,
+    Stdfix,
```
error: Variable not defined: 'Stdfix'
    Stdfix,

https://lab.llvm.org/buildbot/#/builders/163/builds/52530

lntue replied:
https://github.com/llvm/llvm-project/pull/84259

lntue commented:
> runtime build for armv6 baremetal target

Can you check if https://github.com/llvm/llvm-project/pull/84365 fixes the build issue?

nickdesaulniers commented:
Orthogonally, once this is fixed, if you have guidance on how you configure the build of llvm libc for this target, we could look into adding such a build into postsubmit (even if we don't have that specific device, we can still test building for it, even until we have emulators in place to test with).

zeroomega commented:
> Orthogonally, once this is fixed, if you have guidance on how you configure the build of llvm libc for this target, we could look into adding such a build into postsubmit (even if we don't have that specific device, we can still test building for it, even until we have emulators in place to test with).

For armv6m barebone target runtime setup, we use configurations at https://github.com/llvm/llvm-project/blob/2bd369b48dbf0bc3128becb7ef8f8a1b82514b87/clang/cmake/caches/Fuchsia-stage2.cmake#L303 . The runtime was built along side with Clang from source. You might need to change the flags a bit if you want to build the runtime stand alone. @petrhosek knows more about this than me.

Builder https://lab.llvm.org/buildbot/#/builders/98 is already a post submit builder and it should have sent out emails when it fails. Though this builder isn't specifically for barebone runtimes, we just happen to build it along side linux and fuchsia runtimes. Are you looking into adding a presubmit builder for armv6m barebone target?
